src/Lean/Meta/ProofBelow.lean

```diff
@@ -0,0 +1,373 @@
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dany Fabian
+-/
+
+import Lean.Util.Constructions
+import Lean.Meta.Transform
+import Lean.Meta.Tactic
+
+namespace Lean.Meta
+
+namespace ProofBelow
+
+/--
+  The context used in the creation of the `ProofBelow` scheme for inductive predicates.
+-/  
+structure Context where
+  motives : Array (Name × Expr)
+  typeInfos : Array InductiveVal
+  proofBelowNames : Array Name
+  headers : Array Expr
+  numParams : Nat
+
+/--
+  Collection of variables used to keep track of the positions of binders in the construction
+  of `ProofBelow` motives and constructors.
+-/
+structure Variables where
+  target : Array Expr
+  indVal : Array Expr
+  params : Array Expr
+  args : Array Expr
+  motives : Array Expr
+  innerType : Expr
+  deriving Inhabited
+  
+/--
+  Collection of variables used to keep track of the local context used in the `brecOn` proof.
+-/
+structure BrecOnVariables where
+  params : Array FVarId
+  motives : Array FVarId
+  indices : Array FVarId
+  witness : FVarId
+  indHyps : Array FVarId
+
+def mkContext (declName : Name) : MetaM Context := do
+  let indVal ← getConstInfoInduct declName
+  let typeInfos ← indVal.all.toArray.mapM getConstInfoInduct
+  let motiveTypes ← typeInfos.mapM motiveType
+  let motives := ←motiveTypes.mapIdxM fun j motive => do
+    (←motiveName motiveTypes j.val, motive)
+  let headers := typeInfos.mapM $ mkHeader motives indVal.numParams
+  return { 
+    motives := motives
+    typeInfos := typeInfos
+    numParams := indVal.numParams
+    headers := ←headers
+    proofBelowNames := ←indVal.all.toArray.mapM (·.appendAfter "ProofBelow") }
+where 
+  motiveName (motiveTypes : Array Expr) (i : Nat) : MetaM Name := 
+    if motiveTypes.size > 1 
+    then mkFreshUserName s!"motive_{i.succ}"
+    else mkFreshUserName "motive"
+
+  mkHeader 
+    (motives : Array (Name × Expr)) 
```

Comment:
We use 4 spaces when indenting the header.

Reply:
done